casper/src/main/resources/AsyncMap4.rho

```diff
@@ -0,0 +1,213 @@
+// Rholang map desiderata: speedy insert & lookup,  no conflicts on lookup, no conflicts on inserts to different keys
+// This implementation: O(log n) insert & lookup.
+// Conflict analysis
+//   Lookup
+//     The cell implementation uses a linear for, and when two processes send
+//     to the same linear for, you get a conflict. When looking up an item when
+//     you don't know if it's in the map or not, it necessarily sends "get" to
+//     the root cell of the map, so lookups of that kind conflict.  When
+//     looking up an item where you know it's in the map, you can use
+//     "fastUnsafeGet" and avoid the root cell.
+//   Insert
+//     As above, inserting necessarily sends a "get" to the root cell, thus
+//     inserts conflict.
+
+new AsyncMap, o(`rho:io:stdout`) in {
+  new MakeCell, ByteArrayToNybbleList, AsyncMapSetter, AsyncMapGetter, powersCh in {
+    powersCh!([1,2,4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,32768, 65536]) |
+    for (@powers <- powersCh) {
+      contract MakeCell(@val, cell) = {
+        for (@cmd, @args <- cell) {
+          match cmd {
+            "get" => {
+              match args {
+                [ret] => {
+                  @ret!(val) |
+                  MakeCell!(val, *cell)
+                }
+                x => { o!(["ERROR: MATCH FAIL. MakeCell.get line 16", x]) }
+              }
+            }
+            "set" => {
+              match args {
+                [newVal, ret] => {
+                  @ret!(Nil) |
+                  MakeCell!(newVal, *cell)
+                }
+                x => { o!(["ERROR: MATCH FAIL. MakeCell.get line 25", x]) }
+              }
+            }
+            "setBit" => {
+              match args {
+                [bit, ret] => {
+                  @ret!(Nil) |
+                  MakeCell!(
+                    // val | (1 << bit)
+                    // Bitwise operators would be really nice to have!
+                    (val % powers.nth(bit)) +
+                      (val / powers.nth(bit + 1)) * powers.nth(bit + 1) +
+                      powers.nth(bit),
+                    *cell
+                  )
+                }
+                x => { o!(["ERROR: MATCH FAIL. MakeCell.get line 41", x]) }
+              }
+            }
+            "die" => { Nil }
+            x => { o!(["ERROR: MATCH FAIL. MakeCell line 45", x]) }
+          }
+        }
+      } |
+
+      contract ByteArrayToNybbleList(@ba, @n, @len, @acc, ret) = {
+        if (n == len) {
+          ret!(acc)
+        } else {
+          ByteArrayToNybbleList!(ba, n+1, len, acc ++ [ ba.nth(n) % 16, ba.nth(n) / 16 ], *ret)
```

Comment:
this most likely deserves a system process

Reply:
I agree.  The bitwise operators also deserve system processes.